### ISSUE TITLE
Interactive sessions (shells, editors) have various problems

### ISSUE THREAD

**Issue by lnkisi** [paramiko/paramiko#302]
I recently used paramiko implements a login service. In the debugging process, I found a few problems.
I used demo.py and interactive.py on github. After login to the remote server. The text content which was pasted into the command line can not be fully displayed until another key input on keyboard.
The up arrow key should show a recent history command, found it display nothing in the actual test. However, press the Enter key, you can execute the command history.
In addition, arrow keys doesn't work in VIM.

**Comment by blade2005** [comment]
Has any further work been done on this?

**Comment by bitprophet** [comment]
@blade Not specifically but that's partly as there's no single easily pinned down problem to solve.

That said I've been poking at closely related things in my (private but soon to be public) work on Fabric 2, so if I encounter actual bugs in Paramiko (i.e. behaviors that work well locally via subprocess but incorrectly via Paramiko) they will certainly get fixed.

**Comment by blade2005** [comment]
Yeah @bitprophet I noticed the bug in Fabric too but traced it back to parmiko. I'm not sure exactly how this could be fixed. 

My thought was is it a problem in handling stdin/stdout and printing to local stdout and to the pipe. The data is written to the pipe connected to the remote terminal but isn't being displayed locally.

Such as in interactive type
echo hi <return>
<up arrow><return>
so the data is being written to the pipe but not locally indicating a possible issue with writing to stdout and reading all data from the stdout from the pipe. I wonder if dropping from 1024 to something less will work

 *brb gonna go try*

**Comment by blade2005** [comment]
Nope. (I dropped the bytes down to 4)
    x = u(chan.recv(1024))

**Comment by hawick** [comment]
I also have this problem , i can't use vim with arrow key if i press it with slepp 2+ seconds one by one, and history not show the lastest 2 command, but it shows if i press down key after up.

**Comment by bitprophet** [comment]
Thanks for the bump, forgot all about this ticket. I fixed this in pyinvoke/invoke#58 last December; the solution is actually about how one reads stdin, and so it's not relevant to Paramiko itself. (I'm using that code with Paramiko in Fabric's alpha/2.0 branch and it works great - nontrivial colorized vim sessions, etc work fine.)

tl;dr you have to use `FIONREAD` to tell how many bytes are ready for reading from stdin, and read that many, instead of just reading a set amount. _Any_ read resets the 'ready for reading' flag, so multibyte input - like control codes - ends up lost, or read 'later', if you only ever read 1 byte per `select`.

**Comment by kaizamm** [comment]
I have the issue ,too. Another,i have one issue.I used demo.py and interactive.py on github. After login to the remote server. when just one command  input  long enough over  one line, the cursor  will not go to new line, still stay the curent line ,and it will coverage the input before

**Comment by kaizamm** [comment]
@bitprophet Thanks, for your fixing, i solved the vim issue.

**Comment by HengyueLi** [comment]
@kaizamm  How do you solve the problem of "cursor will not go to new line". I confront the same problem.

**Comment by MoAlaaElden** [comment]
> I recently used paramiko implements a login service. In the debugging process, I found a few problems. I used demo.py and interactive.py on github. After login to the remote server. The text content which was pasted into the command line can not be fully displayed until another key input on keyboard. The up arrow key should show a recent history command, found it display nothing in the actual test. However, press the Enter key, you can execute the command history. In addition, arrow keys doesn't work in VIM.

@lnkisi @bitprophet 

Hello All,

Regarding to this issue "The up arrow key should show a recent history command, found it display nothing in the actual test."
I know it was an old issue, but i am able to solve it by adding below lines to interactive.py as shown below:

import os

            if sys.stdin in r:
                #x = sys.stdin.read(1)
                x = os.read(sys.stdin.fileno(), 1) # My code to solve UP Arrow key
                if len(x) == 0:
                    break
                #chan.send(x)
                chan.sendall(x) # My code to solve UP Arrow key
